patches/tModLoader/Terraria/ModLoader/Config/ConfigManager.cs

```diff
@@ -356,6 +459,119 @@ public static bool EnumerableEquals(IEnumerable a, IEnumerable b)
 		}
 		return !hasNextA && !hasNextB;
 	}
+
+	internal static string FormatTextAttribute(string key, string localization, object[] args)
+	{
+		if (args == null)
+			return localization;
+		for (int i = 0; i < args.Length; i++) {
+			if (args[i] is string s && s.StartsWith("$"))
+				args[i] = Language.GetTextValue(FindKeyInScope(s.Substring(1), key));
+		}
+		return Language.GetText(key).Format(args);
+
+		string FindKeyInScope(string key, string scope)
+		{
+			if (LanguageManager.Instance.Exists(key))
+				return key;
+
+			string[] splitKey = scope.Split(".");
+			for (int j = splitKey.Length - 1; j >= 0; j--) {
+				string partialKey = string.Join(".", splitKey.Take(j + 1));
+				string combinedKey = partialKey + "." + key;
+				if (LanguageManager.Instance.Exists(combinedKey))
+					return combinedKey;
+			}
+
+			return key;
+		}
+	}
+
+	private static T GetAndValidate<T>(MemberInfo memberInfo) where T : ConfigKeyAttribute
+	{
+		var configKeyAttribute = (T)Attribute.GetCustomAttribute(memberInfo, typeof(T));
+		if (configKeyAttribute != null) {
+			if (configKeyAttribute.malformed)
+				throw new ValueNotTranslationKeyException($"{nameof(T)} only accepts localization keys for the 'key' parameter.", errorOnType: memberInfo is Type);
+		}
+		return configKeyAttribute;
+	}
+
+	// Used to determine which key to register, based only on field/property, not class, not necessarily which key to use in UI.
+	internal static string GetConfigKey<T>(MemberInfo memberInfo, string dataName) where T : ConfigKeyAttribute
+	{
+		var configKeyAttribute = GetAndValidate<T>(memberInfo);
+		if (configKeyAttribute != null) {
+			return configKeyAttribute.key;
+		}
+
+		// Autokey: Determine key from the Type the member belongs to.
+		return GetDefaultLocalizationKey(memberInfo, dataName);
+	}
+
+	internal static string GetDefaultLocalizationKey(MemberInfo member, string dataName)
+	{
+		Assembly asm = (member is Type t ? t : member.DeclaringType).Assembly;
+		string groupKey = AssemblyManager.GetAssemblyOwner(asm, out var modName) ? $"Mods.{modName}.Configs" : "Config";
+		string memberKey = member is Type ? member.Name : $"{member.DeclaringType.Name}.{member.Name}";
+		return $"{groupKey}.{memberKey}.{dataName}";
+	}
+
+	internal static string GetLocalizedText<T, TArgs>(PropertyFieldWrapper memberInfo, string dataName, bool defaultToNull) where T : ConfigKeyAttribute where TArgs : ConfigArgsAttribute
+	{
+		// Priority: Provided/AutoKey on member -> Provided/AutoKey on class if member translation is empty string -> member name or null
+		var args = (TArgs)Attribute.GetCustomAttribute(memberInfo.MemberInfo, typeof(TArgs));
+		string configKey = GetConfigKey<T>(memberInfo.MemberInfo, dataName: dataName);
+		if (configKey != null && Language.Exists(configKey)) {
```

Comment:
Suggest `#nullable enable` for this file. `configKey` here cannot be null. There's also no way for `Language.Exists` to fail (because we register all of them). This means you can go straight to `Language.GetTextValue`

Reply:
Language.Exists can fail, for example a vanilla class without a registered localization. (EquipType, for one.)